### PR TITLE
Add index_iterator and use it in Perceptron

### PR DESCRIPTION
**Pull request by lisitsyn**

### CODE REVIEW

File: src/shogun/classifier/Perceptron.cpp [hunk 1]
```diff
@@ -3,16 +3,15 @@
  * it under the terms of the GNU General Public License as published by
  * the Free Software Foundation; either version 3 of the License, or
  * (at your option) any later version.
- *
- * Written (W) 1999-2009 Soeren Sonnenburg
- * Copyright (C) 1999-2009 Fraunhofer Institute FIRST and Max-Planck-Society
+ * * Written (W) 1999-2009 Soeren Sonnenburg * Copyright (C) 1999-2009 Fraunhofer Institute FIRST and Max-Planck-Society
```
Comment (karlnapf) [inline src/shogun/classifier/Perceptron.cpp]: gnaaa

File: src/shogun/classifier/Perceptron.cpp [hunk 2]
```diff
@@ -70,15 +71,15 @@ bool CPerceptron::train_machine(CFeatures* data)
...
+			output[example_idx] = features->dense_dot(example_idx, w.vector, w.vlen) + bias;
 
-			if (CMath::sign<float64_t>(output[i]) != train_labels.vector[i])
+			if (CMath::sign<float64_t>(output[example_idx]) != train_labels.vector[example_idx])
 			{
-				converged=false;
-				bias+=learn_rate*train_labels.vector[i];
-				features->add_to_dense_vec(learn_rate*train_labels.vector[i], i, w.vector, w.vlen);
+				converged = false;
+				bias += learn_rate*train_labels.vector[example_idx];
```
Comment (karlnapf) [inline src/shogun/classifier/Perceptron.cpp]: no need for `.vector`

File: src/shogun/features/Features.h [hunk 3]
```diff
@@ -109,6 +110,13 @@ class CFeatures : public CSGObject
 		 */
 		virtual EFeatureClass get_feature_class() const=0;
 
+#ifndef SWIG
+		virtual Range<int32_t> index_iterator() const
```
Comment (karlnapf) [inline src/shogun/features/Features.h]: docs!

**Comment by lisitsyn** [comment]
@karlnapf huh!

File: src/shogun/features/Features.h [hunk 4]
```diff
@@ -109,6 +110,22 @@ class CFeatures : public CSGObject
 		 */
 		virtual EFeatureClass get_feature_class() const=0;
 
+#ifndef SWIG
+		/** returns an iterator of indices
+		 * from 0 to @ref CFeatures::get_num_vectors
+		 *
+		 * Should be used in algorithms in the following way:
+		 * @code
```
Comment (karlnapf) [inline src/shogun/features/Features.h]: muy bien

**Comment by karlnapf** [comment]
open an issue and reference this pr! useful for later

**Pull request merged**